### PR TITLE
UCP/AM: added checking for ucp_am_recv_data_nbx() uses the correct data descriptor

### PR DESCRIPTION
**Pull request by karasevb**
## What
Adds check when the user returns the wrong `UCS_OK` instead of `UCS_INPROGRESS` from AM callback and tries to obtain the data by `ucp_am_recv_data_nbx()`.

### CODE REVIEW

File: src/ucp/core/ucp_am.c [hunk 1]
```diff
@@ -1021,6 +1021,14 @@ UCS_PROFILE_FUNC(ucs_status_ptr_t, ucp_am_recv_data_nbx,
     ucs_status_t status;
     size_t recv_length;
 
+    if (ENABLE_PARAMS_CHECK &&
+        !ucs_unlikely((desc->flags & UCP_RECV_DESC_FLAG_UCT_DESC) ||
+                      (desc->flags & UCP_RECV_DESC_FLAG_RNDV))) {
+        ucs_error("attempt to data receiving for the wrong descriptor");
+        ret = UCS_STATUS_PTR(UCS_ERR_INVALID_PARAM);
+        goto out;
```
Comment (brminich) [inline src/ucp/core/ucp_am.c:1029]: please either return here immdeiately, or move this check after `UCP_WORKER_THREAD_CS_ENTER_CONDITIONAL(worker);`, 
because going to `out` section assumes releasing of worker lock.

File: src/ucp/core/ucp_am.c [hunk 2]
```diff
@@ -1021,6 +1021,14 @@ UCS_PROFILE_FUNC(ucs_status_ptr_t, ucp_am_recv_data_nbx,
     ucs_status_t status;
     size_t recv_length;
 
+    if (ENABLE_PARAMS_CHECK &&
+        !ucs_unlikely((desc->flags & UCP_RECV_DESC_FLAG_UCT_DESC) ||
+                      (desc->flags & UCP_RECV_DESC_FLAG_RNDV))) {
+        ucs_error("attempt to data receiving for the wrong descriptor");
```
Comment (brminich) [inline src/ucp/core/ucp_am.c:1027]: i'd rephrase like
`attempt to receive AM data with invalid descriptor`

File: src/ucp/core/ucp_request.inl [hunk 3]
```diff
@@ -693,9 +693,14 @@ static UCS_F_ALWAYS_INLINE void
 ucp_recv_desc_release(ucp_recv_desc_t *rdesc)
 {
     void *uct_desc;
+    uint16_t flags = rdesc->flags;
+
+    /* reset flags to determine if the descriptor will be wrongly used
+     * after release */
+    rdesc->flags = 0;
```
Comment (brminich) [inline src/ucp/core/ucp_request.inl:700]: let's do it for AM RTS handling only, to avoid performance impact for other flows (tag, stream, am eager, etc)

File: test/gtest/ucp/test_ucp_am.cc [hunk 4]
```diff
@@ -1098,6 +1115,38 @@ UCS_TEST_P(test_ucp_am_nbx_rndv, header_and_data_rndv, "RNDV_THRESH=128")
     test_am_send_recv(127, 1);
 }
 
+UCS_TEST_P(test_ucp_am_nbx_rndv, wrong_rndv_cb, "RNDV_THRESH=1")
+{
+    void *data_desc = NULL;
+    void *rx_data = NULL;
+    char data = 'd';
```
Comment (brminich) [inline test/gtest/ucp/test_ucp_am.cc:1122]: minor: can align by `=`

File: test/gtest/ucp/test_ucp_am.cc [hunk 5]
```diff
@@ -1055,6 +1055,23 @@ class test_ucp_am_nbx_rndv : public test_ucp_am_nbx {
...
 
+    static ucs_status_t am_data_wrong_rndv_cb(void *arg,
+                                              const void *header,
+                                              size_t header_length,
+                                              void *data, size_t length,
+                                              const ucp_am_recv_param_t *param)
+    {
+        void **args = reinterpret_cast<void**>(arg);
+        test_ucp_am_nbx_rndv *self = reinterpret_cast<test_ucp_am_nbx_rndv*>(args[0]);
+        void **data_desc_p = reinterpret_cast<void**>(args[1]);
```
Comment (brminich) [inline test/gtest/ucp/test_ucp_am.cc:1066]: minor: can align by `=`

File: test/gtest/ucp/test_ucp_am.cc [hunk 6]
```diff
@@ -1098,6 +1115,38 @@ UCS_TEST_P(test_ucp_am_nbx_rndv, header_and_data_rndv, "RNDV_THRESH=128")
...
+
+    wait_for_flag(&m_am_received);
+
+    scoped_log_handler wrap_err(wrap_errors_logger);
+    ucs_status_ptr_t rptr = ucp_am_recv_data_nbx(receiver().worker(),
+                                                 data_desc,
+                                                 rx_data, sizeof(data),
+                                                 &param);
+
+    EXPECT_EQ(UCS_ERR_INVALID_PARAM, UCS_PTR_STATUS(rptr));
```
Comment (brminich) [inline test/gtest/ucp/test_ucp_am.cc:1145]: 👍

File: test/gtest/ucp/test_ucp_am.cc [hunk 7]
```diff
@@ -1055,6 +1055,23 @@ class test_ucp_am_nbx_rndv : public test_ucp_am_nbx {
...
+                                              const ucp_am_recv_param_t *param)
+    {
+        void **args = reinterpret_cast<void**>(arg);
+        test_ucp_am_nbx_rndv *self = reinterpret_cast<test_ucp_am_nbx_rndv*>(args[0]);
+        void **data_desc_p = reinterpret_cast<void**>(args[1]);
+
+        *data_desc_p = data;
+        self->m_am_received = true;
+
+        /* return the wrong UCS_OK, instead of UCS_INPROGRESS */
```
Comment (brminich) [inline test/gtest/ucp/test_ucp_am.cc:1071]: please mention that we drop the message, like
"return UCS_OK without calling ucp_am_recv_data_nbx to drop the message"

File: src/ucp/core/ucp_request.inl [hunk 8]
```diff
@@ -693,18 +693,16 @@ static UCS_F_ALWAYS_INLINE void
...
     ucs_trace_req("release receive descriptor %p", rdesc);
-    if (ucs_unlikely(flags & UCP_RECV_DESC_FLAG_UCT_DESC)) {
+    if (ucs_unlikely(rdesc->flags & UCP_RECV_DESC_FLAG_UCT_DESC)) {
         /* uct desc is slowpath */
         uct_desc = UCS_PTR_BYTE_OFFSET(rdesc, -rdesc->uct_desc_offset);
         uct_iface_release_desc(uct_desc);
     } else {
+        /* reset flags to determine if the descriptor will be wrongly used
+         * after release */
+        rdesc->flags = 0;
```
Comment (brminich) [inline src/ucp/core/ucp_request.inl:705]: please add this to AM RTS handler only

File: test/gtest/ucp/test_ucp_am.cc [hunk 9]
```diff
@@ -1098,6 +1116,39 @@ UCS_TEST_P(test_ucp_am_nbx_rndv, header_and_data_rndv, "RNDV_THRESH=128")
     test_am_send_recv(127, 1);
 }
 
+UCS_TEST_P(test_ucp_am_nbx_rndv, wrong_rndv_cb, "RNDV_THRESH=1")
```
Comment (brminich) [inline test/gtest/ucp/test_ucp_am.cc:1119]: let's skip it on valgrind(because it fails with it)
can use
```
UCS_TEST_SKIP_COND_P(test_ucp_am_nbx_rndv, wrong_rndv_cb, RUNNING_ON_VALGRIND)
```
Also please rename the test to `invalid_recv_desc`

**Comment by karasevb** [comment]
@yosefe please review

File: src/ucp/core/ucp_am.c [hunk 10]
```diff
@@ -1558,10 +1565,17 @@ ucs_status_t ucp_am_rndv_process_rts(void *arg, void *data, size_t length,
...
-    if ((desc != NULL) && !(desc->flags & UCP_RECV_DESC_FLAG_UCT_DESC)) {
-        /* Release descriptor if it was allocated on UCP mpool, otherwise it
-         * will be freed by UCT, when UCS_OK is returned from this func. */
-        ucp_recv_desc_release(desc);
+    if (desc != NULL) {
+        /* Specifying the descriptor as released. This can detect the use of
+         * the invalid descriptor in the case when the user returns UCS_OK from
+         * the AM callback and then wrongly tries to receive data with
+         * ucp_am_recv_data_nbx(). */
+        desc->flags |= UCP_RECV_DESC_FLAG_RELEASED;
```
Comment (yosefe) [inline src/ucp/core/ucp_am.c:1573]: do it only if ENABLE_PARAMS_CHECK

File: test/gtest/ucp/test_ucp_am.cc [hunk 11]
```diff
@@ -1098,6 +1116,40 @@ UCS_TEST_P(test_ucp_am_nbx_rndv, header_and_data_rndv, "RNDV_THRESH=128")
...
 }
 
+UCS_TEST_SKIP_COND_P(test_ucp_am_nbx_rndv, invalid_recv_desc,
+                     RUNNING_ON_VALGRIND, "RNDV_THRESH=1")
+{
+    void *data_desc = NULL;
+    void *rx_data   = NULL;
+    char data       = 'd';
+    ucp_request_param_t param;
+    void *args[2];
```
Comment (yosefe) [inline test/gtest/ucp/test_ucp_am.cc:1126]: better use struct, not array

File: src/ucp/core/ucp_am.c (hunk 10 above)
Comment (karasevb) [reply src/ucp/core/ucp_am.c]: fixed

File: test/gtest/ucp/test_ucp_am.cc (hunk 11 above)
Comment (karasevb) [reply test/gtest/ucp/test_ucp_am.cc]: fixed

**Review by brminich — APPROVED** [review]

File: src/ucp/core/ucp_am.c [hunk 12]
```diff
@@ -1566,11 +1566,13 @@ ucs_status_t ucp_am_rndv_process_rts(void *arg, void *data, size_t length,
 
 out:
     if (desc != NULL) {
-        /* Specifying the descriptor as released. This can detect the use of
-         * the invalid descriptor in the case when the user returns UCS_OK from
-         * the AM callback and then wrongly tries to receive data with
-         * ucp_am_recv_data_nbx(). */
-        desc->flags |= UCP_RECV_DESC_FLAG_RELEASED;
+        if(ENABLE_PARAMS_CHECK) {
```
[the remark this answers is not in the document]
Comment (karasevb) [reply src/ucp/core/ucp_am.c]: fixed

File: test/gtest/ucp/test_ucp_am.cc [hunk 13]
```diff
@@ -1006,6 +1006,11 @@ UCP_INSTANTIATE_TEST_CASE(test_ucp_am_nbx_dts)
 
 class test_ucp_am_nbx_rndv : public test_ucp_am_nbx {
 public:
+    struct am_cb_args {
+        void *self;
```
[the remark this answers is not in the document]
Comment (karasevb) [reply test/gtest/ucp/test_ucp_am.cc]: fixed

File: test/gtest/ucp/test_ucp_am.cc [hunk 14]
```diff
@@ -1006,6 +1006,11 @@ UCP_INSTANTIATE_TEST_CASE(test_ucp_am_nbx_dts)
...
+        void *desc;
```
[the remark this answers is not in the document]
Comment (karasevb) [reply test/gtest/ucp/test_ucp_am.cc]: fixed

File: test/gtest/ucp/test_ucp_am.cc [hunk 15]
```diff
@@ -1123,12 +1128,9 @@ UCS_TEST_SKIP_COND_P(test_ucp_am_nbx_rndv, invalid_recv_desc,
     void *rx_data   = NULL;
     char data       = 'd';
     ucp_request_param_t param;
-    void *args[2];
-
-    args[0] = this;
-    args[1] = &data_desc;
+    struct am_cb_args args = { this,  &data_desc };
```
[the remark this answers is not in the document]
Comment (karasevb) [reply test/gtest/ucp/test_ucp_am.cc]: fixed

**Comment by karasevb** [comment]
bot:pipe:retest

File: test/gtest/ucp/test_ucp_am.cc [hunk 16]
```diff
@@ -1007,8 +1007,8 @@ UCP_INSTANTIATE_TEST_CASE(test_ucp_am_nbx_dts)
 class test_ucp_am_nbx_rndv : public test_ucp_am_nbx {
 public:
     struct am_cb_args {
-        void *self;
-        void *desc;
+        test_ucp_am_nbx_rndv *self;
+        void **desc;
```
Comment (yosefe) [inline test/gtest/ucp/test_ucp_am.cc:1011]: indent on column

Comment (karasevb) [reply test/gtest/ucp/test_ucp_am.cc]: Fixed

**Review by yosefe — APPROVED** [review]

**Comment by karasevb** [comment]
Squashed and rebased

**Review by yosefe — APPROVED** [review]

**Comment by yosefe** [comment]
bot:pipe:retest

**Comment by karasevb** [comment]
@yosefe should it ported to 1.10.x branch?
